File: 0261-graph-valid-tree/0261-graph-valid-tree.py
class Solution:
    def validTree(self, n: int, edges: List[List[int]]) -> bool:
        
        par = [-1 for _ in range(n)]
        
        def find(node):
            if par[node] == -1:
                return node
            
            return find(par[node])

        for edge in edges:
            u,v = find(edge[0]), find(edge[1])

            if u != v:
                par[v] = u
            else:
                return False
        
        return len(edges) == n-1

File: 0261-graph-valid-tree/0261-graph-valid-tree.cpp
class Solution {
public:
    bool validTree(int n, vector<vector<int>>& edges) {
        vector<vector<int>> adjList(n);

        if(edges.size() != n-1){
            return false;
        }

        for(const auto &e : edges){
            adjList[e[0]].push_back(e[1]);
            adjList[e[1]].push_back(e[0]);
        }

        vector<bool> visited(n, false);
        stack<int> st;
        st.push(0);

        while(!st.empty()){
            int node = st.top();
            st.pop();
            if(visited[node] == false){
                visited[node] = true;
                for (int neigh : adjList[node]) {
                    if (!visited[neigh])
                        st.push(neigh);
                }
            }
        }
        for(bool v : visited){
            if(!v) {
                return false;
            }
        }
        return true;
    }
};